src/main/java/at/hannibal2/skyhanni/features/itemabilities/CrownOfAvariceCounter.kt

```diff
@@ -24,28 +34,105 @@ object CrownOfAvariceCounter {
     private val internalName = "CROWN_OF_AVARICE".asInternalName()
 
     private var render: Renderable? = null
+    private const val MAX_AVARICE = 1_000_000_000L
+    private val MAX_AFK_TIME = 2.minutes
+
+    private val isWearingCrown by RecalculatingValue(1.seconds) {
+        InventoryUtils.getHelmet()?.getInternalNameOrNull() == internalName
+    }
+
+    private var count: Long = 0L
+    private var coinsEarned: Long = 0L
+    private var sessionStart = SimpleTimeMark.farPast()
+    private var lastCoinUpdate = SimpleTimeMark.farPast()
 
     @SubscribeEvent
-    fun onOverlay(event: GuiRenderEvent.GuiOverlayRenderEvent) {
+    fun onRenderOverlay(event: GuiRenderEvent.GuiOverlayRenderEvent) {
+        if (!isEnabled()) return
+        if (!isWearingCrown) return
         render?.let { config.position.renderRenderable(it, posLabel = "Crown of Avarice Counter") }
     }
 
     @SubscribeEvent
-    fun onTick(event: LorenzTickEvent) {
-        render = check()
+    fun onSecondPassed(event: SecondPassedEvent) {
+        if (!isEnabled()) return
+        if (!isWearingCrown) return
+        update()
+    }
+
+    @SubscribeEvent
+    fun onInventoryUpdated(event: OwnInventoryItemUpdateEvent) {
+        if (!isEnabled() || event.slot != 5) return
+        val item = event.itemStack
+        if (item.getInternalNameOrNull() != internalName) return
+        val coins = item.getCoinsOfAvarice() ?: return
+        if (count == 0L) count = coins
+
+        val coinsDifference = coins - count
+
+        if (coinsDifference == 0L) return
+
+        if (coinsDifference < 0) {
+            reset()
+            count = coins
+            return
+        }
+
+        if (lastCoinUpdate.passedSince() > MAX_AFK_TIME) reset()
+        lastCoinUpdate = SimpleTimeMark.now()
+        coinsEarned += coinsDifference
+        count = coins
+
+        update()
+    }
+
+    @SubscribeEvent
+    fun onWorldChange(event: LorenzWorldChangeEvent) {
+        reset()
+    }
+
+    @SubscribeEvent
+    fun onIslandChange(event: IslandChangeEvent) {
+        reset()
+        count = InventoryUtils.getHelmet()?.getCoinsOfAvarice() ?: return
     }
```

Comment:
I dont feel that both of these are needed

Reply:
> I agree one should be removed but not both, how it originally went was onWorldChange would reset the session i.e. clear coins earned etc. onIslandChange was to prevent a case where someone logs in with CoA already in their helmet slot, showing 0 in the counter than the actual counter, clip to show what i mean : [streamable.com/aney8d](https://streamable.com/aney8d)

i personally dont have the helmet but from the code it looks like you could just use the island change event as that resets it then also sets the count or does this also have issues

Reply:
I think it's like world change but it's fired once you are in the new world which seems ideal for what you are doing. As the world change is sent more during the loading screen